prsa/signal_preprocessing.py

```diff
@@ -0,0 +1,154 @@
+
+    Parameters:
+    time_series (np.ndarray): The input time series, potentially with NaN values.
+
+    Returns:
+    np.ndarray: The normalized time series with values between 0 and 1.
+    """
+    time_series_min = np.nanmin(time_series)
+    time_series_max = np.nanmax(time_series)
+    normalized_time_series = (time_series - time_series_min) / (time_series_max - time_series_min)
+    return normalized_time_series.astype(float)
+
+def detect_peaks(time_series: np.ndarray, filter: np.ndarray, threshold: float = 0.3) -> list[np.ndarray, np.ndarray]:
+    """
+    Detect peaks in a time series using template matching and thresholding.
+
+    Parameters:
+    time_series (np.ndarray): The input time series, potentially with NaN values.
+    filter (np.ndarray): The template used for matching within the time series.
+    threshold (float): The threshold value for peak detection after normalization.
+
+    Returns:
+    Tuple[np.ndarray, np.ndarray]: A tuple containing an array of detected peak indices and the normalized similarity array.
+    """
+    time_series = (time_series - np.nanmean(time_series)) / np.nanstd(time_series)
+    similarity = np.correlate(time_series, filter, mode="same")
+    similarity = similarity / np.max(similarity)
+
+    peaks = np.where(similarity > threshold)[0]
+    return peaks, similarity.astype(float)
+
+def group_peaks(peaks: np.ndarray, threshold: int = 5) -> np.ndarray:
+    """
+    Group nearby peaks within a specified threshold and return the median index of each group.
+
+    Parameters:
+    peaks (np.ndarray): An array of peak indices.
+    threshold (int): The distance threshold for grouping peaks.
+
+    Returns:
+    np.ndarray: An array of the median index for each group of peaks.
+    """
+    output = np.empty(0)
+
+    peak_groups, num_groups = label(np.diff(peaks) < threshold)
+
+    for i in np.unique(peak_groups)[1:]:
+        peak_group = peaks[np.where(peak_groups == i)]
+        output = np.append(output, np.median(peak_group))
+    return output.astype(int)
+
+def extract_rr_intervals_template_matching(time_series: np.ndarray, sampling_rate: int = 200, threshold: float = 0.45, correction_value: int = 2, correction: bool = False) -> list[np.ndarray, np.ndarray]:
+    """
+    Calculate RR intervals from an ECG time series using template matching.
+
+    Description.
+
+    Parameters:
+    time_series (np.ndarray): The input time series, potentially with NaN values.
+    sampling_rate (int): The sampling rate of the ECG signal.
+    threshold (float): The threshold value for peak detection in template matching.
+    correction_value (int): The Z-score value beyond which RR intervals are considered outliers.
+    correction (bool): Flag to apply correction for outlier RR intervals.
+
+    Returns:
+    Tuple[np.ndarray, np.ndarray]: A tuple containing an array of RR intervals and the array of indices of the detected peaks.
+    """
+    # ABP filter
```

Comment:
I am not sure but if this part is written by Ignacy then he needs to mentioned in the repository - please contact him to clarify this